src/metrics.ts

```diff
@@ -0,0 +1,378 @@
+        const requester = new grpc.RequesterBuilder()
+            .withStart((metadata, _listener, next) => {
+                const newListener = new grpc.ListenerBuilder()
+                    .withOnReceiveStatus((status, next) => {
+                        try {
+                            GRPCMetrics.handled({
+                                ...labels,
+                                code: grpc.status[status.code],
+                            });
+                        } finally {
+                            next(status);
+                        }
+                    })
+                    .withOnReceiveMessage((message, next) => {
+                        try {
+                            GRPCMetrics.received(labels);
+                        } finally {
+                            next(message);
+                        }
+                    })
+                    .build();
+                try {
+                    GRPCMetrics.started(labels);
+                } finally {
+                    next(metadata, newListener);
+                }
+            })
+            .withSendMessage((message, next) => {
+                try {
+                    GRPCMetrics.sent(labels);
+                } finally {
+                    next(message);
+                }
+            })
+            .build();
+        return new grpc.InterceptingCall(nextCall(options), requester);
+    };
+}
+
+export class MetricsReporter {
+    private static readonly REPORT_INTERVAL = 60000;
+
+    private metricsHost: string;
+    private intervalHandler: NodeJS.Timer | undefined;
+
+    constructor(gitpodHost: string, private logger: vscode.LogOutputChannel) {
+        const serviceUrl = new URL(gitpodHost);
+        this.metricsHost = `ide.${serviceUrl.hostname}`;
+    }
+
+    startReporting() {
+        if (this.intervalHandler) {
+            return;
+        }
+        this.intervalHandler = setInterval(() => this.report().catch(e => this.logger.error('Error while reporting metrics', e)), MetricsReporter.REPORT_INTERVAL);
+    }
+
+    private async report() {
+        const metrics = await register.getMetricsAsJSON();
+        register.resetMetrics();
+        for (const m of metrics) {
+            const type = m.type as unknown as string;
+            if (type === 'counter') {
+                await this.reportCounter(m);
+            } else if (type === 'histogram') {
+                await this.reportHistogram(m);
+            }
+        }
+    }
+
+    private async reportCounter(metric: metric) {
+        const counterMetric = metric as metric & { values: [{ value: number; labels: Record<string, string> }] };
+        for (const { value, labels } of counterMetric.values) {
+            if (value > 0) {
+                await this.doAddCounter(counterMetric.name, labels, value);
+            }
+        }
+    }
+
+    private async reportHistogram(metric: metric) {
```

Comment:
For `reportHistogram` in supervisor/reporter.go we need previous value to get diff value. But here is nothing related to previous value, should not we copy logic in supervisor and translate it into js (maybe with the help of copilot lab xd)?

Reply:
No need to calculate difference as I call `register.resetMetrics();` above, and because this are just counters internally is the same as diff

Reply:
Woo, it's cool that we can do it that simple way in nodejs 👍